File: reto9/reto9.ts
type Space = "·" | "@" | "*" | "o";
type Board = Space[];
type Movement = "U" | "D" | "R" | "L";
type Result = "none" | "crash" | "eat";

function moveTrain(board: Board, mov: Movement): Result {
  const y = board.findIndex((row) => row.includes("@"));
  const x = board[y].indexOf("@");

  const moves = { U: [-1, 0], D: [1, 0], L: [0, -1], R: [0, 1] };
  const [dy, dx] = moves[mov];
  const [ny, nx] = [y + dy, x + dx];

  if (ny < 0 || ny >= board.length || nx < 0 || nx >= board[0].length)
    return "crash";

  const symbol: Record<Space, Result> = {
    o: "crash",
    "*": "eat",
    "@": "crash",
    "·": "none",
  };
  return symbol[board[ny][nx]];
}

File: reto9/reto9.py
from typing import List, Literal


def move_train(
    board: List[str], mov: Literal["U", "D", "R", "L"]
) -> Literal["none", "crash", "eat"]:
    y = [i for i, row in enumerate(board) if "@" in row][0]
    x = board[y].index("@")

    dy = {"U": -1, "D": 1, "R": 0, "L": 0}[mov]
    dx = {"U": 0, "D": 0, "R": 1, "L": -1}[mov]
    ny, nx = y + dy, x + dx

    if not (0 <= ny < len(board) and 0 <= nx < len(board[0])):
        return "crash"
    return {"o": "crash", "*": "eat"}.get(board[ny][nx], "none")
